main/client/src/mill/main/client/FileToStreamTailer.java

```diff
@@ -0,0 +1,97 @@
+package mill.main.client;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.Optional;
+
+public class FileToStreamTailer extends Thread implements AutoCloseable {
+
+	private final File file;
+	private final PrintStream stream;
+	private final int intervalMsec;
+
+	private Optional<BufferedReader> reader = Optional.empty();
+	// if true, we won't read the whole file, but only new lines
+	private boolean ignoreHead = true;
+
+	private volatile boolean keepReading = true;
+	private volatile boolean flush = false;
+
+	public FileToStreamTailer(File file, PrintStream stream, int intervalMsec) {
+		super("Tail");
+		this.intervalMsec = intervalMsec;
+		setDaemon(true);
+		this.file = file;
+		this.stream = stream;
+	}
+
+	@Override
+	public void run() {
+		if (isInterrupted()) {
+			keepReading = false;
+		}
+		while (keepReading || flush) {
+			flush = false;
+			try {
+				if (!reader.isPresent()) {
+					try {
+						this.reader = Optional.of(new BufferedReader(new FileReader(file)));
+					} catch (FileNotFoundException e) {
+						// nothing to ignore if file is inially missing
+						ignoreHead = false;
+					}
+				}
+				reader.ifPresent(r -> {
+					// read lines
+					try {
+						String line;
+						while ((line = r.readLine()) != null) {
+							if (!ignoreHead) {
+								stream.println(line);
+							}
+						}
+						// we ignored once
+						this.ignoreHead = false;
+					} catch (FileNotFoundException e) {
+						// File vanished
```

Comment:
In what scenarios would the "File is missing" and "File vanished" errors occur? I wonder if we can specify the problem away, e.g. by having the protocol be that someone *must* create the log files before these tailers are started, and the server waits until the client has finished tailing before picking up another client. Otherwise it seems hard to be confident we handle all scenarios where the log file an disappear from under you

Reply:
There is no scenario in mill besides "the user deleted it". I will just drop the catch.